---
title: "Functional enrichment analysis"
output: html_document
author: "J. Ignacio Lucas Lled√≥"
params:
   TAG: 'genes'
   VAR: 'regime'
bibliography: enrichment.bib
---

```{r setup, message=FALSE}
library(topGO)
library(knitr)   # to use kable()
library(limma)   # to use vennDiagram()
library(ggplot2)
library(variancePartition)
TAG <- params$TAG
VAR <- params$VAR
ANNOTATION <- list(genes = '../2019-07-26/genes/annotation.txt',
                   isoforms = '../2019-07-26/transcripts/annotation.txt')
DGEDIR <- '../2020-01-08/'
ENRICHMENT <- paste('../2020-01-14/Enrichment', TAG, VAR, 'RData', sep='.')
```

# Introduction
This is the enrichment analysis for `r TAG` regulated by `r VAR`. I am using the 
`topGO` package [@Alexa2019], which is able to apply different algorithms. The starting point
is an ordering of `r TAG` usually according to their association with a phenotype or
any other relevant quantity, such as a measure of differential expression between
two conditions. In this case, we order `r TAG` by their differential expression
between `r VAR` levels. Following one of the suggestions in the manual, I originally
ordered `r TAG` by either $p$ value of the differential expression analysis, or by
the (complement of the) amount of its expression variance explained by `r VAR`
(see `2020-01-14`). Those were quite equivalent orderings. However, they do not
distinguish the sign of the fold change: both up- and down-regulated genes contribute
to make a gene ontology term significant. It is one of the approaches recommended,
and it makes sense, to the extent that among all the genes annotated with a function,
some may display contrasting expression patterns, for example if both positive and
negative regulators of that function are annotated with the same label. I thus,
expect the use of $p$ values in enrichment analysis to facilitate the detection of
high level categories. 

In contrast, genes sharing lower level categories (very specific GO terms) 
are more likely to be expressed in a similar way. Then, an ordering of genes that
reflect the sign of the difference in expression levels between the two conditions
compared would be more informative. In principle, enrichment methods should be able
to detect an enrichment in either end of the list, rather than only in the top.
However, I noticed that the Kolmogorov-Smirnov test implemented in `topGO` only
looks at the *top* of the list. Even if the score can be negative, it treats it
as if it always was a *p* value, and orders genes from lowest to highest. Even if
the function that selects for *significant* genes uses the absolute value of the
score, the Kolmogorov-Smirnov test does not use any threshold, and ignores that
detail.

In any case, I don't want to use a different statistic based on a pre-selected
subset of *significant* genes. In addition to that being an arbitrary decision, the
*ordering* of genes would become folded again. I just need to run the enrichment
analysis with the KS statistic twice: once with the *t* statistic, and once with
its reverse, $-t$. The first test will find GO terms overrepresented among the genes
that are overexpressed in the regular, relative to the random environment. The second
test will find GO terms overrepresented among genes overexpressed in the random
environment.

# Reading the data
Functional annotation is in `2019-07-26`. I will also upload two lists of `r TAG`,
with either proportion of variance explained by `r VAR` or p-value of differential
expression test.

```{r data}
load(paste0(DGEDIR, TAG, '.RData'))
tagTStat    <- topTable(fitmm, coef = VAR, num = length(fitmm$F), sort.by='t', resort.by='t')[,'t',drop=FALSE]
Annotation  <- read.table(ANNOTATION[[TAG]], col.names = c('tagname', 'goterms'))
```

To initialize the `topGOdata` object, I need the gene list as a named numeric vector,
where the names are the `r TAG` identifiers and the numeric values, the $t$ statistics
from the differential expression analysis. The `structure()` function adds attributes
to an object.

```{r scores}
TStats <- structure(tagTStat[,1], names = row.names(tagTStat))
TMinus <- -TStats
rm(tagTStat)
```

There are `r length(TStats)` `r TAG` scored with a $t$ statistic.
The `Annotation` data frame has more than one GO term for every tag, separated by `|`.
I need a named list.

```{r annotation}
head(Annotation)
allgenes2GO <- strsplit(as.character(Annotation$goterms), "|", fixed = TRUE)
names(allgenes2GO) <- Annotation$tagname
rm(Annotation)
```

There are `r length(allgenes2GO)` `r TAG` with GO annotations. But the differential expression
analysis includes many more `r TAG`. In order to include the not-annotated `r TAG` in the
enrichment analysis, to see if annotated or not annotated `r TAG` are more or less often
differentially expressed, I should attribute a GO term to them. According to [http://geneontology.org/docs/faq/]
nowadays we express lack of annotation by annotating to the root nodes, i.e. GO:0008150 biological_process,
GO:0003674 molecular_function, and GO:0005575 cellular_component.

```{r unknown}
for (tag in names(TStats)) {
   if (! tag %in% names(allgenes2GO)) {
      allgenes2GO <- append(allgenes2GO,
         structure(list(c("GO:0008150", "GO:0003674", "GO:0005575")), names = tag))
   }
}
```

# Building the topGO objects

Creation of a topGO dataset is documented in section 4 of topGO's the user manual:
https://bioconductor.org/packages/release/bioc/vignettes/topGO/inst/doc/topGO.pdf.
I need to use the `new`
command, and fill up the slots. The `annot` object must be a function that compiles "a list of
GO terms such that each element in the list is a character vector containing all the gene 
identifiers that are mapped to the respective GO term." There are several options, that you can
check using `help(annFUN.gene2GO)`, for example. The `annFUN.gene2GO` requires a `gene2GO`
argument, which is the list of gene-to-GO terms I made before. The `geneSelectionFun` object is a
function that selects the *interesting* (most significant) genes. It is required to
perform Fisher's exact test. The `nodeSize` is used to prune the GO hierarchy from the terms
which have less than `n` annotated genes.

I generate three datasets, to analyse each of the three ontologies. These are the
three topGO objects for the regular regime.

```{r GOdata1, results = 'hold', message = FALSE}
# with 'ks' statistics, the selection function only affects the display of
# presumed numbers of 'significant' genes, I think.
selection <- function(allScores) {return(allScores < -5.5)}
GOdata.BP <- new('topGOdata',
   description = 'Differential expression among Brachionus plicatilis populations.',
   ontology = 'BP', 
   allGenes = TStats,
   annot = annFUN.gene2GO,
   gene2GO = allgenes2GO,
   geneSelectionFun = selection,
   nodeSize = 5)

GOdata.MF <- new('topGOdata',
   description = 'Differential expression among Brachionus plicatilis populations.',
   ontology = 'MF',
   allGenes = TStats,
   annot = annFUN.gene2GO,
   gene2GO = allgenes2GO,
   geneSelectionFun = selection,
   nodeSize = 5)

GOdata.CC <- new('topGOdata',
   description = 'Differential expression among Brachionus plicatilis populations.',
   ontology = 'CC',
   allGenes = TStats,
   annot = annFUN.gene2GO,
   gene2GO = allgenes2GO,
   geneSelectionFun = selection,
   nodeSize = 5)

DataSummary <- data.frame(ontology = c('BP', 'MF', 'CC'),
   Num_Genes = sapply(list(GOdata.BP, GOdata.MF, GOdata.CC), numGenes),
   Num_GO_terms = sapply(list(GOdata.BP, GOdata.MF, GOdata.CC), function(x) length(usedGO(x))))
kable(DataSummary, caption='Number of feasible genes or transcripts and number of GO terms used in each data set.')
rm(DataSummary)
```

And the following three, for the random regime.

```{r GOdata2, results = 'hold', message = FALSE}
GOminus.BP <- new('topGOdata',
   description = 'Differential expression among Brachionus plicatilis populations.',
   ontology = 'BP', 
   allGenes = TMinus,
   annot = annFUN.gene2GO,
   gene2GO = allgenes2GO,
   geneSelectionFun = selection,
   nodeSize = 5)

GOminus.MF <- new('topGOdata',
   description = 'Differential expression among Brachionus plicatilis populations.',
   ontology = 'MF',
   allGenes = TMinus,
   annot = annFUN.gene2GO,
   gene2GO = allgenes2GO,
   geneSelectionFun = selection,
   nodeSize = 5)

GOminus.CC <- new('topGOdata',
   description = 'Differential expression among Brachionus plicatilis populations.',
   ontology = 'CC',
   allGenes = TMinus,
   annot = annFUN.gene2GO,
   gene2GO = allgenes2GO,
   geneSelectionFun = selection,
   nodeSize = 5)
```

# Running the tests

There are more than one way to test for enrichment. Something that took me a while to understand
is that not only there are different test statistics (Fisher's exact test, Kolmogorov-Smirnov,
and others) but also different *algorithms*: classic, elim, weight... The algorithms are ways
to deal with the dependence structure among GO terms due to topology. Some algorithms are
compatible with all statistics implemented in topGO. But the *weight* and the *parentchild*
algorithms are only applicable to Fisher's exact test. I am not interested in the classic
algorithm, which treats GO nodes as independent, and therefore produces an excess of significant
results. I will not use the Fisher's exact test, because it dependes on an arbitrary threshold
of significance on non-adjusted p-values.

I start testing for enrichment of GO terms among genes overexpressed in the regular
environment.

```{r testing1, message = FALSE}
BP.reg.elim     <- runTest(GOdata.BP, algorithm = "elim",     statistic = "ks")
BP.reg.weight01 <- runTest(GOdata.BP, algorithm = "weight01", statistic = "ks")
BP.reg.lea      <- runTest(GOdata.BP, algorithm = "lea",      statistic = "ks")
MF.reg.elim     <- runTest(GOdata.MF, algorithm = "elim",     statistic = "ks")
MF.reg.weight01 <- runTest(GOdata.MF, algorithm = "weight01", statistic = "ks")
MF.reg.lea      <- runTest(GOdata.MF, algorithm = "lea",      statistic = "ks")
CC.reg.elim     <- runTest(GOdata.CC, algorithm = "elim",     statistic = "ks")
CC.reg.weight01 <- runTest(GOdata.CC, algorithm = "weight01", statistic = "ks")
CC.reg.lea      <- runTest(GOdata.CC, algorithm = "lea",      statistic = "ks")

ResultsSummary <- data.frame(ontology = rep(c("BP", "MF", "CC"), each = 3),
   algorithm = rep(c("elim", "weight01", "lea"), 3),
   TermsTested = sapply(list(BP.reg.elim, BP.reg.weight01, BP.reg.lea,
                             MF.reg.elim, MF.reg.weight01, MF.reg.lea,
                             CC.reg.elim, CC.reg.weight01, CC.reg.lea), function(X) length(score(X))),
   Significant = sapply(list(BP.reg.elim, BP.reg.weight01, BP.reg.lea,
                             MF.reg.elim, MF.reg.weight01, MF.reg.lea,
                             CC.reg.elim, CC.reg.weight01, CC.reg.lea), function(X) sum(score(X) < 0.01)))

kable(ResultsSummary, caption="Number of non-trivial terms tested and those with a score (not corrected p-value) lower than 0.01.")
rm(ResultsSummary)
```

And now in the random environment:

```{r testing2, message = FALSE}
BP.ran.elim     <- runTest(GOminus.BP, algorithm = "elim",     statistic = "ks")
BP.ran.weight01 <- runTest(GOminus.BP, algorithm = "weight01", statistic = "ks")
BP.ran.lea      <- runTest(GOminus.BP, algorithm = "lea",      statistic = "ks")
MF.ran.elim     <- runTest(GOminus.MF, algorithm = "elim",     statistic = "ks")
MF.ran.weight01 <- runTest(GOminus.MF, algorithm = "weight01", statistic = "ks")
MF.ran.lea      <- runTest(GOminus.MF, algorithm = "lea",      statistic = "ks")
CC.ran.elim     <- runTest(GOminus.CC, algorithm = "elim",     statistic = "ks")
CC.ran.weight01 <- runTest(GOminus.CC, algorithm = "weight01", statistic = "ks")
CC.ran.lea      <- runTest(GOminus.CC, algorithm = "lea",      statistic = "ks")

ResultsSummary <- data.frame(ontology = rep(c("BP", "MF", "CC"), each = 3),
   algorithm = rep(c("elim", "weight01", "lea"), 3),
   TermsTested = sapply(list(BP.ran.elim, BP.ran.weight01, BP.ran.lea,
                             MF.ran.elim, MF.ran.weight01, MF.ran.lea,
                             CC.ran.elim, CC.ran.weight01, CC.ran.lea), function(X) length(score(X))),
   Significant = sapply(list(BP.ran.elim, BP.ran.weight01, BP.ran.lea,
                             MF.ran.elim, MF.ran.weight01, MF.ran.lea,
                             CC.ran.elim, CC.ran.weight01, CC.ran.lea), function(X) sum(score(X) < 0.01)))

kable(ResultsSummary, caption="Number of non-trivial terms tested and those with a score (not corrected p-value) lower than 0.01.")
rm(ResultsSummary)
```

# Functional enrichment in predictable environment

## Biological process

```{r BP1, message=FALSE}
orderedTerms <- names(sort(score(BP.reg.weight01)))
significant.weight01 <- score(BP.reg.weight01)[orderedTerms] <= 0.005
significant.lea      <- score(BP.reg.lea)[orderedTerms] <= 0.005
significant.elim     <- score(BP.reg.elim)[orderedTerms] <= 0.005
sigTerms <- orderedTerms[significant.weight01 & significant.lea & significant.elim]
# sigTerms gets updated, and is already used in the code, but I need a permanent copy for later.
BP.reg.sigTerms <- sigTerms

BP.all <- GenTable(GOdata.BP, elim=BP.reg.elim, weight01=BP.reg.weight01, lea=BP.reg.lea,
                   orderBy="weight01", ranksOf="elim", topNodes=sum(significant.elim))

kable(
   BP.all[BP.all$GO.ID %in% sigTerms,],
   caption = "Most overrepresented Biological Process terms in predictable environments")

vennDiagram(vennCounts(cbind(weight01=significant.weight01,
                             lea=significant.lea,
                             elim=significant.elim)))

kable(data.frame(Term = Term(GOTERM[sigTerms]),
                 Definition = Definition(GOTERM[sigTerms]),
                 PValue=score(BP.reg.weight01)[sigTerms]),
      caption = 'BP terms overrepresented in predictable environments')
```

I think the GO graph is useful to see the relationship among the significant terms. But
too large graphs are impossible to read. I don't know how to split the graph in meaningful
subgraphs.

```{r BP1dag, message=FALSE, fig.width=30, fig.height=15}
showSigOfNodes(GOdata.BP, score(BP.reg.weight01),
               firstSigNodes = sum(significant.elim),
               wantedNodes = sigTerms)
```

This is just a example of the most significant GO term:

```{r BP1density, message=FALSE}
showGroupDensity(GOdata.BP, orderedTerms[1], rm.one=FALSE)
showGroupDensity(GOdata.BP, orderedTerms[2], rm.one=FALSE)
showGroupDensity(GOdata.BP, orderedTerms[3], rm.one=FALSE)
```

## Molecular function

```{r MF1, message=FALSE}
orderedTerms <- names(sort(score(MF.reg.weight01)))
significant.weight01 <- score(MF.reg.weight01)[orderedTerms] <= 0.005
significant.lea      <- score(MF.reg.lea)[orderedTerms] <= 0.005
significant.elim     <- score(MF.reg.elim)[orderedTerms] <= 0.005
sigTerms <- orderedTerms[significant.weight01 & significant.lea & significant.elim]
# sigTerms gets updated, and is already used in the code, but I need a permanent copy for later.
MF.reg.sigTerms <- sigTerms

MF.all <- GenTable(GOdata.MF, elim=MF.reg.elim, weight01=MF.reg.weight01, lea=MF.reg.lea,
                   orderBy="weight01", ranksOf="elim", topNodes=sum(significant.elim))

kable(
   MF.all[MF.all$GO.ID %in% sigTerms,],
   caption = "Most overrepresented Molecular Function terms in predictable environments.")

vennDiagram(vennCounts(cbind(weight01=significant.weight01,
                             lea=significant.lea,
                             elim=significant.elim)))

kable(data.frame(Term=Term(GOTERM[sigTerms]),
                 Definition=Definition(GOTERM[sigTerms]),
                 PValue=score(MF.reg.weight01)[sigTerms]),
      caption = 'MF terms overrepresented in predictable environments')
```

```{r MF1dag, message=FALSE, fig.width=30, fig.height=15}
showSigOfNodes(GOdata.MF, score(MF.reg.weight01),
               firstSigNodes = sum(significant.elim),
               wantedNodes = sigTerms)
```

```{r MF1density, message=FALSE}
showGroupDensity(GOdata.MF, orderedTerms[1], rm.one=FALSE)
showGroupDensity(GOdata.MF, orderedTerms[2], rm.one=FALSE)
showGroupDensity(GOdata.MF, orderedTerms[3], rm.one=FALSE)
```

## Cellular component

```{r CC1, message=FALSE}
orderedTerms <- names(sort(score(CC.reg.weight01)))
significant.weight01 <- score(CC.reg.weight01)[orderedTerms] <= 0.005
significant.lea      <- score(CC.reg.lea)[orderedTerms] <= 0.005
significant.elim     <- score(CC.reg.elim)[orderedTerms] <= 0.005
sigTerms <- orderedTerms[significant.weight01 & significant.lea & significant.elim]
# sigTerms gets updated, and is already used in the code, but I need a permanent copy for later.
CC.reg.sigTerms <- sigTerms

CC.all <- GenTable(GOdata.CC, elim=CC.reg.elim, weight01=CC.reg.weight01, lea=CC.reg.lea,
                   orderBy="weight01", ranksOf="elim", topNodes=sum(significant.elim))
kable(
   CC.all[CC.all$GO.ID %in% sigTerms,],
   caption = "Most overrepresented Cellular Component terms in predictable environments")

vennDiagram(vennCounts(cbind(weight01=significant.weight01,
                             lea=significant.lea,
                             elim=significant.elim)))

kable(data.frame(Term=Term(GOTERM[sigTerms]),
                 Definition=Definition(GOTERM[sigTerms]),
                 PValue=score(CC.reg.weight01)[sigTerms]),
      caption = 'CC terms overrepresented in predictable environments')
```

```{r CC1dag, message=FALSE, fig.width=30, fig.height=15}
showSigOfNodes(GOdata.CC, score(CC.reg.weight01),
               firstSigNodes = sum(significant.elim),
               wantedNodes = sigTerms)
```

```{r CC1density, message=FALSE}
showGroupDensity(GOdata.CC, orderedTerms[1], rm.one=FALSE)
showGroupDensity(GOdata.CC, orderedTerms[2], rm.one=FALSE)
showGroupDensity(GOdata.CC, orderedTerms[3], rm.one=FALSE)
```

# Functional enrichment in the unpredictable environment

## Biological process

```{r BP2, message=FALSE}
orderedTerms <- names(sort(score(BP.ran.weight01)))
significant.weight01 <- score(BP.ran.weight01)[orderedTerms] <= 0.005
significant.lea      <- score(BP.ran.lea)[orderedTerms] <= 0.005
significant.elim     <- score(BP.ran.elim)[orderedTerms] <= 0.005
sigTerms <- orderedTerms[significant.weight01 & significant.lea & significant.elim]
# sigTerms gets updated, and is already used in the code, but I need a permanent copy for later.
BP.ran.sigTerms <- sigTerms

BP.all <- GenTable(GOminus.BP, elim=BP.ran.elim, weight01=BP.ran.weight01, lea=BP.ran.lea,
                   orderBy="weight01", ranksOf="elim", topNodes=sum(significant.elim))

kable(
   BP.all[BP.all$GO.ID %in% sigTerms,],
   caption = "Most overrepresented Biological Process terms in random environments.")

vennDiagram(vennCounts(cbind(weight01=significant.weight01,
                             lea=significant.lea,
                             elim=significant.elim)))

kable(data.frame(Term = Term(GOTERM[sigTerms]),
                 Definition = Definition(GOTERM[sigTerms]),
                 PValue=score(BP.ran.weight01)[sigTerms]),
      caption = 'BP terms overrepresented in random environments')
```

I think the GO graph is useful to see the relationship among the significant terms. But
too large graphs are impossible to read. I don't know how to split the graph in meaningful
subgraphs.

```{r BP2dag, message=FALSE, fig.width=30, fig.height=15}
showSigOfNodes(GOminus.BP, score(BP.ran.weight01),
               firstSigNodes = sum(significant.elim),
               wantedNodes = sigTerms)
```

This is just a example of the most significant GO term:

```{r BP2density, message=FALSE}
showGroupDensity(GOminus.BP, orderedTerms[1], rm.one=FALSE)
showGroupDensity(GOminus.BP, orderedTerms[2], rm.one=FALSE)
showGroupDensity(GOminus.BP, orderedTerms[3], rm.one=FALSE)
```

## Molecular function

```{r MF2, message=FALSE}
orderedTerms <- names(sort(score(MF.ran.weight01)))
significant.weight01 <- score(MF.ran.weight01)[orderedTerms] <= 0.005
significant.lea      <- score(MF.ran.lea)[orderedTerms] <= 0.005
significant.elim     <- score(MF.ran.elim)[orderedTerms] <= 0.005
sigTerms <- orderedTerms[significant.weight01 & significant.lea & significant.elim]
# sigTerms gets updated, and is already used in the code, but I need a permanent copy for later.
MF.ran.sigTerms <- sigTerms

MF.all <- GenTable(GOminus.MF, elim=MF.ran.elim, weight01=MF.ran.weight01, lea=MF.ran.lea,
                   orderBy="weight01", ranksOf="elim", topNodes=sum(significant.elim))

kable(
   MF.all[MF.all$GO.ID %in% sigTerms,],
   caption = "Most significant terms of the Molecular Function ontology.")

vennDiagram(vennCounts(cbind(weight01=significant.weight01,
                             lea=significant.lea,
                             elim=significant.elim)))

kable(data.frame(Term=Term(GOTERM[sigTerms]),
                 Definition=Definition(GOTERM[sigTerms]),
                 PValue=score(MF.ran.weight01)[sigTerms]),
      caption = 'MF terms overrepresented in random environments')
```

```{r MF2dag, message=FALSE, fig.width=30, fig.height=15}
showSigOfNodes(GOminus.MF, score(MF.ran.weight01),
               firstSigNodes = sum(significant.elim),
               wantedNodes = sigTerms)
```

```{r MF2density, message=FALSE}
showGroupDensity(GOminus.MF, orderedTerms[1], rm.one=FALSE)
showGroupDensity(GOminus.MF, orderedTerms[2], rm.one=FALSE)
showGroupDensity(GOminus.MF, orderedTerms[3], rm.one=FALSE)
```

## Cellular component

```{r CC2, message=FALSE}
orderedTerms <- names(sort(score(CC.ran.weight01)))
significant.weight01 <- score(CC.ran.weight01)[orderedTerms] <= 0.005
significant.lea      <- score(CC.ran.lea)[orderedTerms] <= 0.005
significant.elim     <- score(CC.ran.elim)[orderedTerms] <= 0.005
sigTerms <- orderedTerms[significant.weight01 & significant.lea & significant.elim]
# sigTerms gets updated, and is already used in the code, but I need a permanent copy for later.
CC.ran.sigTerms <- sigTerms

CC.all <- GenTable(GOminus.CC, elim=CC.ran.elim, weight01=CC.ran.weight01, lea=CC.ran.lea,
                   orderBy="weight01", ranksOf="elim", topNodes=sum(significant.elim))
kable(
   CC.all[CC.all$GO.ID %in% sigTerms,],
   caption = "Most overrepresented Cellular Component terms in random environments.")

vennDiagram(vennCounts(cbind(weight01=significant.weight01,
                             lea=significant.lea,
                             elim=significant.elim)))

kable(data.frame(Term=Term(GOTERM[sigTerms]),
                 Definition=Definition(GOTERM[sigTerms]),
                 PValue=score(CC.ran.weight01)[sigTerms]),
      caption = 'CC terms overrepresented in random environments.')
```

```{r CC2dag, message=FALSE, fig.width=30, fig.height=15}
showSigOfNodes(GOminus.CC, score(CC.ran.weight01),
               firstSigNodes = sum(significant.elim),
               wantedNodes = sigTerms)
```

```{r CC2density, message=FALSE}
showGroupDensity(GOminus.CC, orderedTerms[1], rm.one=FALSE)
showGroupDensity(GOminus.CC, orderedTerms[2], rm.one=FALSE)
showGroupDensity(GOminus.CC, orderedTerms[3], rm.one=FALSE)
```

# Comparison between the two ordering of genes

I want to compare these results with those from `2020-01-14`, where I used
the $p$ value of the differential expression analysis to order the genes.
I can import the data from `2020-01-14`, but I need to do it in a new
environment to prevent overwriting the current results.

```{r newEnv}
load(ENRICHMENT, ex <- new.env())
```

## Biological process

```{r BPcomparison, message=FALSE}
allTerms <- usedGO(GOdata.BP)
BP.pvalue.sigTerms <- with(ex, BP.pvalue.sigTerms)
vennDiagram(vennCounts(cbind(TStat  = allTerms %in% c(BP.reg.sigTerms, BP.ran.sigTerms),
                             PValue = allTerms %in% BP.pvalue.sigTerms)))
# New terms:
as.data.frame(Term(GOTERM[setdiff(c(BP.reg.sigTerms, BP.ran.sigTerms), BP.pvalue.sigTerms)]))

# Absent terms:
as.data.frame(Term(GOTERM[setdiff(BP.pvalue.sigTerms, c(BP.reg.sigTerms, BP.ran.sigTerms))]))
```

## Molecular function

```{r MFcomparison, message=FALSE}
allTerms <- usedGO(GOdata.MF)
MF.pvalue.sigTerms <- with(ex, MF.pvalue.sigTerms)
vennDiagram(vennCounts(cbind(TStat  = allTerms %in% c(MF.reg.sigTerms, MF.ran.sigTerms),
                             PValue = allTerms %in% MF.pvalue.sigTerms)))
# New terms:
as.data.frame(Term(GOTERM[setdiff(c(MF.reg.sigTerms, MF.ran.sigTerms), MF.pvalue.sigTerms)]))

# Absent terms:
as.data.frame(Term(GOTERM[setdiff(MF.pvalue.sigTerms, c(MF.reg.sigTerms, MF.ran.sigTerms))]))
```

## Cellular component

```{r CCcomparison, message=FALSE}
allTerms <- usedGO(GOdata.CC)
CC.pvalue.sigTerms <- with(ex, CC.pvalue.sigTerms)
vennDiagram(vennCounts(cbind(TStat  = allTerms %in% c(CC.reg.sigTerms, CC.ran.sigTerms),
                             PValue = allTerms %in% CC.pvalue.sigTerms)))
# New terms:
as.data.frame(Term(GOTERM[setdiff(c(CC.reg.sigTerms, CC.ran.sigTerms), CC.pvalue.sigTerms)]))

# Absent terms:
as.data.frame(Term(GOTERM[setdiff(CC.pvalue.sigTerms, c(CC.reg.sigTerms, CC.ran.sigTerms))]))
```

# Session info

```{r last}
save(allgenes2GO,
     GOdata.BP,  BP.reg.elim, BP.reg.weight01, BP.reg.lea, BP.reg.sigTerms,
     GOminus.BP, BP.ran.elim, BP.ran.weight01, BP.ran.lea, BP.ran.sigTerms,
     GOdata.MF,  MF.reg.elim, MF.reg.weight01, MF.reg.lea, MF.reg.sigTerms,
     GOminus.MF, MF.ran.elim, MF.ran.weight01, MF.ran.lea, MF.ran.sigTerms,
     GOdata.CC,  CC.reg.elim, CC.reg.weight01, CC.reg.lea, CC.reg.sigTerms,
     GOminus.CC, CC.ran.elim, CC.ran.weight01, CC.ran.lea, CC.ran.sigTerms,
     file = paste('Enrichment', TAG, VAR, 'RData', sep='.'))
sessionInfo()
```