---
title: "Figures"
author: "J. Ignacio Lucas Lled√≥"
date: "29/6/2020"
output: html_document
---

```{r setup, include=TRUE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(variancePartition)
library(tidyr)
library(GO.db)
library(stringr)
library(topGO)
library(grid)
library(gridExtra)
varPart.results      <- '../2020-01-08/genes.RData'
annotation.results   <- '../2019-07-26/genes/annotation.txt'
interproscan.results <- '../2019-07-19/functiona_annotation.tsv'
trans2gene.results   <- '../2020-01-23/transcript2gene.txt'
GO.results <- '../2020-01-14/Enrichment.genes.regime.RData'
```


# Variance explained (Figure 4)

The `plotVarPart()` function does not let you customize the plot. I need to use `ggplot()`
to change the order of factors. I also change their names to match.

```{r Fig4}
load(varPart.results)
fig4data <- as.data.frame(varPart)
fig4data <- pivot_longer(fig4data, 1:5, names_to = 'modelTerm', values_to = 'varExplained')
fig4data$modelTerm <- factor(fig4data$modelTerm,
                             levels = c('regime','treatment','regime:treatment','population','Residuals'),
                             labels = c('Selective regime', 'Diapause condition', 'Sel. regime x D. condition',
                                        'Population(Regime)', 'Residuals'))
fig4data$varExplained <- 100 * fig4data$varExplained
ggplot(data = fig4data, mapping = aes(x = modelTerm, y = varExplained)) + theme_bw() +
  xlab('') + ylab('Variance explained (%)') +
  theme(axis.text.x = element_text(angle = 20, color = 'black', hjust = 1), aspect.ratio = 0.7) +
  geom_violin(scale = 'width', aes(fill = modelTerm), show.legend = FALSE) + 
  geom_boxplot(fill = 'gray', width=0.1) + scale_fill_manual(values = c(ggColorHue(4), 'grey85'))
ggsave('Fig4.png')
rm(fig4data)
```

# Volcano plots (Figures 5 and 6)

I need to join several plots together and use the same colors in all plots.

```{r volcano}
annotation <- read.table(annotation.results,
                         col.names = c('gene', 'GOterms'),
                         colClasses = c('character', 'character'))
annotation.list <- strsplit(annotation$GOterms, split='|', fixed=TRUE)
names(annotation.list) <- annotation$gene

# append can only join two vectors at a time. It works with lists!
allAncestors <- append(as.list(GOBPANCESTOR),
                       append(as.list(GOMFANCESTOR),
                              as.list(GOCCANCESTOR)))

fullAnnotation <- lapply(annotation.list,
                         FUN = function(x){
                           unique(append(x,
                                         unlist(allAncestors[x], use.names=FALSE)))
                         })

getGenes <- function(goterm, annotation = fullAnnotation, varpart = varPart) {
  genes <- names(annotation[grep(paste(goterm, collapse='|'), annotation)])
  if (! is.null(varpart)) {
    genes <- genes[genes %in% row.names(varpart)]
  }
  return(genes)
}

volcanoPlot <- function(fit=fitmm, Coef='regime', num=length(fitmm$F),
                        term=goterm, genes=getGenes(term)){
  DE <- topTable(fit, coef=Coef, number=num)
  DE$annotation <- factor('other', levels=c(Term(GOTERM[term]), 'other'))
  DE[row.names(DE) %in% genes, 'annotation'] <- Term(GOTERM[term])
  DE <- DE[order(DE$annotation, decreasing=TRUE),]
  ggplot(data=DE, mapping=aes(x=logFC, y=-log10(P.Value), color=annotation)) +
    geom_point(show.legend = FALSE) +
    ggtitle(str_wrap(ifelse(nchar(Term(GOTERM[term])) <= 70,
                            paste0(term, ", ", Term(GOTERM[term])),
                            paste0(term, ", ", substr(Term(GOTERM[term]), 1, 60), '...')),
                     width=40)) +
    #xlab(expression(log[2]('Fold Change'))) +
    #ylab(expression(-log[10]('p-value'))) +
    xlab('') + ylab('') +
    scale_color_manual(values = ggColorHue(2))
}
```

```{r Fig5, fig.width=12, fig.height=8}
diapause_maintenance <- c('GO:0006289', 'GO:0006979', 'GO:0016715',
                          'GO:0005992', 'GO:0007186', 'GO:0007165')
p <- grid.arrange(grobs = lapply(diapause_maintenance, function(x) volcanoPlot(term=x)),
             left = grid.text(expression(-log[10]('p value')), rot = 90, gp = gpar(fontsize = 18)),
             bottom = grid.text(expression(log[2]('Fold Change')), gp = gpar(fontsize = 18)),
             top = grid.text('Functions associated with diapause maintenance', gp = gpar(fontsize = 18)),
             nrow = 2)
p
ggsave('Fig5.png', plot=p)
```

```{r Fig6, fig.width=12, fig.height=12}
embryo_development   <- c('GO:0007160', 'GO:0006813', 'GO:0046942',
                          'GO:0048870', 'GO:0003341', 'GO:0060271',
                          'GO:0004181', 'GO:0004252', 'GO:0004222')
p <- grid.arrange(grobs = lapply(embryo_development, function(x) volcanoPlot(term=x)),
             left = grid.text(expression(-log[10]('p value')), rot = 90, gp = gpar(fontsize = 18)),
             bottom = grid.text(expression(log[2]('Fold Change')), gp = gpar(fontsize = 18)),
             top = grid.text('Functions associated with embryo development', gp = gpar(fontsize = 18)),
             nrow = 3)
p
ggsave('Fig6.png', plot=p)

```

# Comparison of resuts of gene-set enrichment analyses (Supplementary figure 2)

Supplementary figure 2 is meant only to show that the enrichment analyses
performed with two alternative gene orderings are equivalent. The two gene
orders correspond to: a) the $p$ value of the differential gene expression analysis
performed in `2020-01-08` with the `dream()` function; and b) the amount of
variance in gene expression explained by the relevant factor (selective regime).
The gene-set enrichment analyses (`2020-01-14`) determine how significantly enriched
gene ontology terms are among the top genes. The scatterplots below, thus, represent
gene ontology terms. I use ranks, instead of $p$ values, because the distribution of
ranks makes the relationship easier to visualize. In any case, the purpose of those
plots is simply to note that the two alternative gene orderings offer very similar
results.

```{r FigS2, fig.height=12, fig.width=4, message=FALSE}
load(GO.results)
allTerms <- usedGO(GOdata.BP)
stopifnot(all.equal(usedGO(GOdata.BP), usedGO(GOdataVar.BP)))
p1 <- ggplot(data = data.frame(DGE     = rank(score(BP.weight01))[allTerms],
                               VarPart = rank(score(BPvar.weight01))[allTerms]),
             mapping = aes(x = DGE, y = VarPart)) + geom_point() +
  geom_smooth(method='lm') + xlab('') +
  ylab('Using portion of explained variance') + ggtitle('Biological Process (GO:0008150)')

allTerms <- usedGO(GOdata.MF)
stopifnot(all.equal(usedGO(GOdata.MF), usedGO(GOdataVar.MF)))
p2 <- ggplot(data = data.frame(DGE     = rank(score(MF.weight01))[allTerms],
                               VarPart = rank(score(MFvar.weight01))[allTerms]),
             mapping = aes( x = DGE, y = VarPart)) + geom_point() +
  geom_smooth(method = 'lm') + xlab('') +
  ylab('Using portion of explained variance') + ggtitle('Molecular Function (GO:0003674)')

allTerms <- usedGO(GOdata.CC)
stopifnot(all.equal(usedGO(GOdata.CC), usedGO(GOdataVar.CC)))
p3 <- ggplot(data = data.frame(DGE     = rank(score(CC.weight01))[allTerms],
                               VarPart = rank(score(CCvar.weight01))[allTerms]),
             mapping = aes( x = DGE, y = VarPart)) + geom_point() +
  geom_smooth(method = 'lm') + xlab('Using differential expression p-values') +
  ylab('Using portion of explained variance') + ggtitle('Cellular Component (GO:0005575)')

p4 <- grid.arrange(p1, p2, p3, ncol=1)
p4
ggsave('FigureS2.png', plot = p4)
```

# Session Information

```{r sessionInfo}
sessionInfo()
```