---
title: "The effect of environmental randomness is population-specific"
author: "Ignasi"
date: "23/12/2019"
output: html_document
---

```{r setup, include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(edgeR)
library(ggplot2)
library(reactable)
library(cp4p)
COUNTS_FILE='../2019-03-29/genes.PostCount.txt'
```

The following is the same as in `../2019-04-03/Preliminar.Rmd`. Because edgeR is fast enough
to fit the model, and because I did not save an environment, I just run this again here. I am
fitting only model 4, because it has already been determined that there are no genes with a
significant interaction between selective regime and hatching condition, but there are significant
interaction terms between selective regime and original population. It seems reasonable that no
interaction between the two main factors can be observed if different populations are responding
differently to the selective regime.

```{r models, cache=TRUE}
# Variables
counts <- read.table(COUNTS_FILE, row.names=1, header=TRUE)
population <- factor(c("1","1","2","2","3","3","4","4","5","5","6","6"))
NestedPop  <- factor(c("1","1","2","2","1","1","3","3","2","2","3","3"))
diapause   <- factor(rep(c("hatch","diapause"), 6))
diapause   <- relevel(diapause, "hatch")
regime     <- factor(c("random","random","random","random","regular","regular",
                 "random","random","regular","regular","regular","regular"))
regime     <- relevel(regime, "regular")

# Filtering
y <- DGEList(counts=counts, group=regime)
y <- calcNormFactors(y, method="TMM")
threshold <- 5.0 / (min(y$samples$lib.size) / 1000000)
keep <- rowSums(cpm(y) > threshold) >= 4
y <- y[keep, ,keep.lib.sizes=FALSE]
NumOfTags <- dim(y)[1]

# Model
design4 <- model.matrix(~regime + diapause + regime:NestedPop)
row.names(design4) <- names(counts)
reactable(design4, columns=list(
  '(Intercept)' = colDef(name='intercept'),
  regimerandom  = colDef(name='random'),
  diapausediapause = colDef(name='diapause'),
  'regimeregular:NestedPop2' = colDef(name='regular:2'),
  'regimeregular:NestedPop3' = colDef(name='regular:3'),
  'regimerandom:NestedPop2'  = colDef(name='random:2'),
  'regimerandom:NestedPop3'  = colDef(name='random:3')))
y4 <- estimateDisp(y, design4)
fit4 <- glmQLFit(y4, design4)
```

Instead of reporting the number of significantly regulated genes under a specific FDR,
I would like to report the overall, FDR-independent, estimated fraction of regulated
genes. I know there are ways to estimate the number of true null hypotheses, even when
there is a dependence structure among tests. I found the package `cp4p` (Gianetto et al.
2016. Proteomics, 16(1), 29-32.), that does exactly that.

```{r tests}
DE.random    <- glmQLFTest(fit4, coef = 2)
DE.regular.2 <- glmQLFTest(fit4, coef = 4)
DE.random.2  <- glmQLFTest(fit4, coef = 5)
DE.regular.3 <- glmQLFTest(fit4, coef = 6)
DE.random.3  <- glmQLFTest(fit4, coef = 7)
PropNonDE <- rbind(estim.pi0(DE.random$table$PValue)$pi0.est,
                   estim.pi0(DE.regular.2$table$PValue)$pi0.est,
                   estim.pi0(DE.random.2$table$PValue)$pi0.est,
                   estim.pi0(DE.regular.3$table$PValue)$pi0.est,
                   estim.pi0(DE.random.3$table$PValue)$pi0.est)
row.names(PropNonDE) <- c('random', 'regular.2', 'random.2', 'regular.3', 'random.3')
reactable(t(PropNonDE), defaultColDef = colDef(format = colFormat(digits = 2)))
```

These are the estimated proportions of genes the expression of which is *not* affected by the
condition represented by the coefficient being tested. Namely, coefficient
2 (*random*) represents the effect of the random selective regime, relative to the regular
regime; coefficient 4 (*regular.2*) is the effect of the second population within the regular
regime (population 5) relative to the first population in the regular regime (population 3);
coefficient 5 (*random.2*) is the effec of the second population in the random regime (population
2) relative to the first population in the random regime (population 1); coefficient 6
(*regular.3*) is the effect of the third population in the regular regime (population 6)
relative to the first population in the regular regime (population 3); and coefficient 7 (*random.3*)
is the effect of the third population in the random regime (population 4) relative to the
first population in the random regime (population 1).

The SLIM method [Want et al. 2011](https://academic.oup.com/bioinformatics/article/27/2/225/286449),
is supposed to be adequate for datasets with dependence structure. I am inclined to believe its results,
even though it suggests that the third population under regular regime has a lot of genes expressed
differently than the first population under regular regime.

```{r calibrationPlots}
calibration.plot(DE.random$table$PValue, pi0.method = 'slim')
calibration.plot(DE.regular.2$table$PValue, pi0.method = 'slim')
calibration.plot(DE.random.2$table$PValue, pi0.method = 'slim')
calibration.plot(DE.regular.3$table$PValue, pi0.method = 'slim')
calibration.plot(DE.random.3$table$PValue, pi0.method = 'slim')
```

But I think it would make more sense to compare each population in a regime to the same baseline.
For example, the interesting question is not how many and what genes are expressed differently between
population 1 and population 2, both under the random regime (coefficient 5). The question is how
many and what genes in population 2 (or 1, or 3) are differentially expressed with respect to all
populations in the regular regime. We can do that using contrasts.