---
title: "Effect of selective regime on gene expression in B. plicatilis"
output: html_document
author: "J. Ignacio Lucas Lled√≥"
---

```{r eval=TRUE, echo=FALSE, include=FALSE}
library(edgeR)
library(ggplot2)
args <- commandArgs(trailingOnly = TRUE)
if (length(args) == 0) {
   stop("At least one argument must be supplied: input file.", call.=FALSE)
} else if (length(args) == 1) {
   args[2] <- "."
}

input_file_name <- args[1]
output_dir <- args[2]

population <- factor(c("1","1","2","2","3","3","4","4","5","5","6","6"))
NestedPop  <- factor(c("1","1","2","2","1","1","3","3","2","2","3","3"))
diapause   <- factor(rep(c("hatch","diapause"), 6))
diapause   <- relevel(diapause, "hatch")
regime     <- factor(c("random","random","random","random","regular","regular",
                 "random","random","regular","regular","regular","regular"))
regime     <- relevel(regime, "regular")
combined   <- factor(c("RandHatch","RandDiap","RandHatch","RandDiap","RegulHatch","RegulDiap",
                      "RandHatch","RandDiap","RegulHatch","RegulDiap","RegulHatch","RegulDiap"))
combined   <- relevel(combined, "RegulHatch")

if (!exists('y')) {
   if (!exists('counts')) {
      counts <- read.table(input_file_name, row.names=1, header=TRUE)
   }
   y <- DGEList(counts=counts, group=regime)
   y <- calcNormFactors(y, method="TMM")
   threshold <- 5.0 / (min(y$samples$lib.size) / 1000000)
   keep <- rowSums(cpm(y) > threshold) >= 4
   y <- y[keep, ,keep.lib.sizes=FALSE]
}

NumOfTags <- dim(y)[1]

if (!exists('fit1')) {
   design1 <- model.matrix(~population + diapause)
   if (!exists('y1')) {
      y1 <- estimateDisp(y, design1)
   }
   fit1 <- glmQLFit(y1, design1)
}

if (!exists('fit2')) {
   design2 <- model.matrix(~regime + diapause)
   if (!exists('y2')) {
      y2 <- estimateDisp(y, design2)
   }
   fit2 <- glmQLFit(y2, design2)   
}

if (!exists('fit3')) {
   design3 <- model.matrix(~0 + combined)
   if (!exists('y3')) {
      y3 <- estimateDisp(y, design3)
   }
   fit3 <- glmQLFit(y3, design3)
}

if (!exists('fit4')) {
   design4 <- model.matrix(~regime + diapause + regime:NestedPop)
   if (!exists('y4')) {
      y4 <- estimateDisp(y, design4)
   }
   fit4 <- glmQLFit(y4, design4)
}

if (!exists('fit5')) {
   design5 <- model.matrix(~regime + diapause + regime:diapause)
   if (!exists('y5')) {
      y5 <- estimateDisp(y, design5)
   }
   fit5 <- glmQLFit(y5, design5)
}

if (!exists('fit6')) {
   design6 <- model.matrix(~regime + diapause + regime:diapause + regime:NestedPop)
   if (!exists('y6')) {
      y6 <- estimateDisp(y, design6)
   }
   fit6 <- glmQLFit(y6, design6)
}

if (!exists('I1vsI2')) {
   I1vsI2 <- read.table("../2017-10-27/list_genes/I1vsI2_all.txt", 
      colClasses=c("NULL", rep("character", 6), rep("numeric", 6),"factor"),
      col.names=c("ignored","geneid","gene","locus","sample1","sample2","testStat",
      "FPKMx","FPKMy","log2FC","statistic","p","q","significant"))
}

if (!exists('P1vsP2')) {
   P1vsP2 <- read.table("../2017-10-27/list_genes/P1vsP2_all.txt",
      colClasses=c("NULL", rep("character", 6), rep("numeric", 6),"factor"),
      col.names=c("ignored","geneid","gene","locus","sample1","sample2","testStat",
      "FPKMx","FPKMy","log2FC","statistic","p","q","significant"))
}

if (!exists('I1vsP1')) {
   I1vsP1 <- read.table("../2017-10-27/list_genes/I1vsP1_all.txt",
      colClasses=c("NULL", rep("character", 6), rep("numeric", 6),"factor"),
      col.names=c("ignored","geneid","gene","locus","sample1","sample2","testStat",
      "FPKMx","FPKMy","log2FC","statistic","p","q","significant"))
}

if (!exists('I2vsP2')) {
   I2vsP2 <- read.table("../2017-10-27/list_genes/I2vsP2_all.txt",
      colClasses=c("NULL", rep("character", 6), rep("numeric", 6),"factor"),
      col.names=c("ignored","geneid","gene","locus","sample1","sample2","testStat",
      "FPKMx","FPKMy","log2FC","statistic","p","q","significant"))
}

if (!exists('Cuffdiff_hatching')) {
   Cuffdiff_hatching <- I1vsI2$geneid[I1vsI2$geneid %in% P1vsP2$geneid]
}

if (!exists('Cuffdiff_regime')) {
   Cuffdiff_regime   <- I1vsP1$geneid[I1vsP1$geneid %in% I2vsP2$geneid]
}
```
# Model 2

```{r}
qlf2  <- glmQLFTest(fit2, coef=2)
topTags(qlf2)
z1 <- topTags(qlf2, n=NumOfTags)$table
z1$logPValue <- -log10(z1$PValue)
ggplot(data=z1, mapping=aes(x=logFC, y=logPValue, color=FDR)) +
   geom_point() + theme_grey() + ylab(expression(paste("-log"[10], plain("p value)")))) +
   ggtitle('Main effect of selective regime')
```

According to model 2, there are `r dim(topTags(qlf2,n=NumOfTags,p.value=0.01))[1]` tags
significantly up- or downregulated by selective regime at FDR=0.01.

# Model 3

Using model 3 I can look for genes affected by selective regime in two ways. First, I could
look for a significant change in average gene expression among samples in the same regime.
Alternatively, I could separately look for genes affected by the selective regime among samples
in the same hatching condition, and then compare the lists of significant genes. According to
Edu, the latter makes more sense biologically, because we expect the selective regime to affect
different sets of genes depending on whether eggs are allowed to hatch or *forced* to undergo
diapause.

## Effect of selective regime when eggs are allowed to hatch immediately

```{r}
qlf3a  <- glmQLFTest(fit3, contrast=c(-1, 0, 1, 0))
topTags(qlf3a)
z1 <- topTags(qlf3a, n=NumOfTags)$table
z1$logPValue <- -log10(z1$PValue)
ggplot(data=z1, mapping=aes(x=logFC, y=logPValue, color=FDR)) +
   geom_point() + theme_grey() + ylab(expression(paste("-log"[10], plain("p value)")))) +
   ggtitle('Effect of selective regime, when hatching')
```
According to model 3, contrast *a*, there are `r dim(topTags(qlf3a, n=NumOfTags, p.value=0.05))[1]` tags
differentially expressed between selective regimes among samples allowed to hatch immediately, at FDR=0.05.

## Effect of selective regime when eggs are forced to undergo diapause

```{r}
qlf3b  <- glmQLFTest(fit3, contrast=c(0, 1, 0, -1))
topTags(qlf3b)
z1 <- topTags(qlf3b, n=NumOfTags)$table
z1$logPValue <- -log10(z1$PValue)
ggplot(data=z1, mapping=aes(x=logFC, y=logPValue, color=FDR)) +
   geom_point() + theme_grey() + ylab(expression(paste("-log"[10], plain("p value)")))) +
   ggtitle('Effect of selective regime, when hatching')
```

According to model 3, contrast *b*, there are `r dim(topTags(qlf3b, n=NumOfTags, p.value=0.05))[1]` tags
differentially expressed between selective regimes among samples undergoing diapause, at FDR=0.05. There is
a reasonable correlation between the fold change in the two hatching conditions:

```{r}
z2 <- merge(topTags(qlf3a, n=NumOfTags)$table,
            topTags(qlf3b, n=NumOfTags)$table,
            by='row.names')
ggplot(data=z2, mapping=aes(x=logFC.x, y=logFC.y, color=FDR.x)) +
   geom_point() + geom_smooth(method='lm') + xlab('logFC hatching') + ylab('logFC diapause')
```

And this is the overlap between genes (or isoforms) affected by selective regime when hatching is immediate and
those affected by selective regime when diapause is induced:

```{r}
significant <- cbind(model3a = topTags(qlf3a, n=NumOfTags, sort.by='none')$table$FDR <= 0.05,
                     model3b = topTags(qlf3b, n=NumOfTags, sort.by='none')$table$FDR <= 0.05)
vennDiagram(vennCounts(significant))
```

## Average effect of selective regime

```{r}
qlf3c  <- glmQLFTest(fit3, contrast=c(-0.5, 0.5, 0.5, -0.5))
topTags(qlf3c)
z1 <- topTags(qlf3c, n=NumOfTags)$table
z1$logPValue <- -log10(z1$PValue)
ggplot(data=z1, mapping=aes(x=logFC, y=logPValue, color=FDR)) +
   geom_point() + theme_grey() + ylab(expression(paste("-log"[10], plain("p value)")))) +
   ggtitle('Effect of selective regime, when hatching')
```

According to model 3, contrast *c*, there are `r dim(topTags(qlf3b, n=NumOfTags, p.value=0.05))[1]` tags
with a significant change (across selective regimes) of average (across hatching conditions) expression
levels, at FDR = 0.05. 

# Model 4

```{r}
qlf4  <- glmQLFTest(fit4, coef=2)
topTags(qlf4)
z1 <- topTags(qlf4, n=NumOfTags)$table
z1$logPValue <- -log10(z1$PValue)
ggplot(data=z1, mapping=aes(x=logFC, y=logPValue, color=FDR)) +
   geom_point() + theme_grey() + ylab(expression(paste("-log"[10], plain("p value)")))) +
   ggtitle('Main effect of selective regime, model 4')
```

According to model 4, there are `r dim(topTags(qlf4,n=NumOfTags,p.value=0.25))[1]` tags
significantly up- or downregulated by selective regime at FDR=0.25. I wrote some of them
to a text file:

```{r}
write.table(topTags(qlf4, n=NumOfTags, p.value=0.15), file=sprintf("%s/regime.txt", args[2]), quote=FALSE)
```

# Model 5

```{r}
qlf5  <- glmQLFTest(fit5, coef=2)
topTags(qlf5)
z1 <- topTags(qlf5, n=NumOfTags)$table
z1$logPValue <- -log10(z1$PValue)
ggplot(data=z1, mapping=aes(x=logFC, y=logPValue, color=FDR)) +
   geom_point() + theme_grey() + ylab(expression(paste("-log"[10], plain("p value)")))) +
   ggtitle('Main effect of selective regime, model 5')
```

According to model 5, there are `r dim(topTags(qlf5,n=NumOfTags,p.value=0.05))[1]` tags
significantly up- or downregulated by selective regime at FDR=0.25.

# Model 6

```{r}
qlf6  <- glmQLFTest(fit6, coef=2)
topTags(qlf6)
z1 <- topTags(qlf6, n=NumOfTags)$table
z1$logPValue <- -log10(z1$PValue)
ggplot(data=z1, mapping=aes(x=logFC, y=logPValue, color=FDR)) +
   geom_point() + theme_grey() + ylab(expression(paste("-log"[10], plain("p value)")))) +
   ggtitle('Main effect of selective regime, model 6')
```

According to model 6, there are `r dim(topTags(qlf6,n=NumOfTags,p.value=0.25))[1]` tags
significantly up- or downregulated by selective regime at FDR=0.25. And this is how the
different models agree on what genes are differentially expressed in different selective
regimes:

```{r}
significant <- cbind(model2 = topTags(qlf2,  n=NumOfTags, sort.by='none')$table$FDR <= 0.01,
                     model3 = topTags(qlf3a, n=NumOfTags, sort.by='none')$table$FDR <= 0.05,
                     model4 = topTags(qlf4,  n=NumOfTags, sort.by='none')$table$FDR <= 0.25,
                     model5 = topTags(qlf5,  n=NumOfTags, sort.by='none')$table$FDR <= 0.05,
                     model6 = topTags(qlf6,  n=NumOfTags, sort.by='none')$table$FDR <= 0.25)
vennDiagram(vennCounts(significant))
```

Note that I am using a different FDR for each model, to obtain more or less comparable
and positive numbers.
