---
title: "Variance partition of gene expression in a split-plot design"
author: "J. Ignacio Lucas Lled√≥"
date: "9/1/2020"
output: html_document
bibliography: mixed.bib
---

# Introduction
The EdgeR package [@Robinson2010] has the advantage of modelling count data, instead of relative expression
levels. It offers some capacity to fit complex models [@McCarthy2012]. But it does not include random effects.
The experiment was a split-plot design, which requires a mixed model with the random effect of
the original population of origin in order to properly account for the correlation of expression
levels in sub-populations of the same origin.

The `variancePartition` package [@Hoffman2016] fits mixed models to the expression data of every gene in the
dataset. It also includes the `dream()` function [Hoffman2020] to test for differential expression under this
framework. It does require the expression level to be normalized, though.

```{r setup, message=FALSE}
library(variancePartition)
library(edgeR)
library(BiocParallel)
library(cp4p)
COUNTS_FILE <- '../2019-03-29/genes.PostCount.txt'
NUM_THREADS <- 40  # Number of cores for parallelization.
TAG <- 'genes'
```

# Variance partition
When fitting the model to every gene's expression profile, information from all genes contribute to
every fit. The shared information helps determine the dispersion of every measure. Counts are heteroscedastic:
their variance depend on their mean. The exact mean-variance relationship may be complex, but it can
be learned from the data. The `limma` package [@Ritchie2015] uses the `voom()` function [@Law2014] to 1) transform count data to
log2-counts per million, and to estimate the mean-variance relationship. Once that relationship is estimated,
every expression level is assigned a *weight*: the inverse of its predicted variance. This weights account for
the heteroscedasticity and are used afterwards in fitting the model and in testing for differential expression.

Curiously, though, a model needs to be fit first, for every gene. The residual standard deviation and the
average gene expression level of that fit is one point in the scatterplot where the mean-variance relationship
becomes apparent. When fitting these gene-wise linear models, `voom()` does not consider random errors, but
`voomWithDreamWeights()` does.

```{r data, cache=TRUE}
counts <- as.matrix(read.table(COUNTS_FILE, row.names=1, header=TRUE))

## Filtering
# A gene is expressed if it has at least 5 reads, which in terms of CPM in the smallest
# library is equal to `threshold`. Only genes expressed in at least 4 samples are retained.
threshold <- 5 / (min(colSums(counts)) / 1000000)
keep <- rowSums(cpm(counts) > threshold) >= 4
geneExpr <- DGEList(counts[keep,])
geneExpr <- calcNormFactors(geneExpr)
numOfTags <- dim(geneExpr)[1]

## Metadata
metadata <- data.frame(row.names = c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5",
                                  "X3A_S9", "X3C_S11", "X4A_S6", "X4C_S12",
                                  "X5A_S2", "X5C_S4", "X6A_S3", "X6C_S10"),
                       regime = c(1, 1, 1, 1, 0, 0 , 1, 1, 0, 0, 0, 0),
                       regimeChar = c("random", "random", "random", "random",
                                  "regular", "regular", "random", "random",
                                  "regular", "regular", "regular", "regular"),
                       population = c("1", "1", "2", "2", "3", "3",
                                      "4", "4", "5", "5", "6", "6"),
                       treatment = rep(c(0, 1), 6),
                       treatmentChar = rep(c("hatch", "diapause"), 6))

## Model (includes interaction term)
form <- ~ regime * treatment + (1|population)
```

The `variancePartition` package requires all categorical variables to be random effects. This is a way to prevent
bias when estimating the variance component of categorical variables with many levels. However, the
`dream()` function seems to require the variable to be tested be a fixed effect. To get out of this catch22,
I dummy-code regime and treatment as numeric variables with values 0 and 1.

```{r fits, cache=TRUE, warning=FALSE, message=FALSE}
## Parallelization
param = SnowParam(NUM_THREADS, "SOCK")
register(param)

## Weights
vobjDream <- voomWithDreamWeights(geneExpr, form, metadata)

## Variance partition
varPart <- fitExtractVarPartModel(vobjDream, form, metadata)
plotVarPart(varPart)
plotPercentBars(varPart[1:20,])
```

Recall that the partition of variance shows relative amounts of variance explained by individual predictors,
irrespectively of the absolute amount of variation. Thus, genes for which a large fraction of their expression
variance is explained by the selective regime do not necessarily have large fold change values. Let's take a
look at the stratified expression level of the gene with largest fraction of variance explained by regime:

```{r stratified}
i <- which.max(varPart$regime)

# Expression level is log2 counts per million.
GE <- data.frame(Expression = vobjDream$E[i,], Regime = metadata$regimeChar)
plotStratify(Expression ~ Regime, GE, main = rownames(vobjDream$E)[i])
rm(GE)
```

Below I write the lists of genes with their fraction of expression variance explained by either selective
regime or hatching condition. The genes are ordered from the one the expression of which is most affected
by that variable to the one that is least affected. Note that among the top genes there will be both positive
and negatively affected genes.

```{r geneLists}
# Ordered by regime effect:
head(varPart[order(varPart$regime, decreasing = TRUE),])
write.table(varPart[order(varPart$regime, decreasing = TRUE),2],
            file = paste(TAG, '/regime_variance.txt', sep=''), quote=FALSE,
            row.names = TRUE, col.names = TRUE)

# Order by hatching condition effect:
head(varPart[order(varPart$treatment, decreasing=TRUE),])
write.table(varPart[order(varPart$treatment, decreasing = TRUE),3],
            file = paste(TAG, '/hatching_variance.txt', sep=''), quote = FALSE,
            row.names = TRUE, col.names = TRUE)
```

# Differential expression

```{r dream, cache=TRUE}
fitmm <- dream( vobjDream, form, metadata )
head(fitmm$p.value)
```

The `estim.pi0` function of the `cp4p` package [@Gianetto2016] estimates the proportion of null hypothesis from a 
list of p values. That is, it estimates the proportion of genes the expression of which is *not*
affected by the corresponding factor.

```{r pi0}
pi0.regime <- estim.pi0(fitmm$p.value[,'regime'])
pi0.regime
pi0.treatment <- estim.pi0(fitmm$p.value[,'treatment'])
pi0.treatment
calibration.plot(fitmm$p.value[,'regime'], pi0.method='abh')
calibration.plot(fitmm$p.value[,'treatment'], pi0.method='abh')
```

The `dream()` function uses a different, estimated number of degrees of freedom for every gene, which
makes t-statistics not directly comparable. The `write.fit()` function reports t-statistics and p-values
for all coefficients. Below, I use `topTable()`, which transforms p-values in a signed z-score, which is usable
for downstream analysis. Note that for the selective regime, a positive z-score (or a positive log-fold change)
means the gene is expressed more in the random than in the regular regime. And for the hatching condition
(or *treatment* coefficient), positive z-score (or log-fold change) means that the gene is expressed more
under forced diapause (samples labelled with the *C* letter) than under immediate hatching.

I doubt whether to separate up- and downregulated genes in opposite ends of the gene list that will be used
in functional enrichment analysis. It is clear that if a functional category contain genes both up- and downregulated,
I may fail to identify that category as significant if I separate them [@Ackermann2009]. For the moment, I will use the
p-value, rather than the z-score, to order genes, ignoring the sign of the effect. This will make the lists
more comparable to those obtained from variance partition.

```{r pvalue}
regime_pvalue <- topTable(fitmm, coef = 'regime', number = numOfTags, sort.by = 'p')
write.table(regime_pvalue[,'P.Value'], file = paste(TAG, '/regime_pvalue.txt', sep=''), quote = FALSE,
            row.names = TRUE, col.names = TRUE)

hatching_pvalue <- topTable(fitmm, coef = 'treatment', number = numOfTags, sort.by = 'p')
write.table(hatching_pvalue[,'P.Value'], file = paste(TAG, '/hatching_pvalue.txt', sep=''), quote = FALSE,
            row.names = TRUE, col.names = TRUE)
```

# Session Info

```{r sessioninfo}
save(vobjDream, varPart, fitmm, file = paste(TAG, '.RData', sep=''))
sessionInfo()
```

# References
