---
title: "Cell functions affected by environmental unpredictability"
author: "J. Ignacio Lucas Lled√≥"
date: "23/1/2020"
output: html_document
bibliography: interpretation.bib
---

```{r hide, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

On `2020-01-14` I run the analysis of Gene Ontology terms. Here I want to focus on part of those results and make some findings
clearer. I will use results only from gene annotation, and not from annotation of transcripts, which is a somewhat noiser dataset.
I will not look at the functions affected by hatching condition, since it only affects a couple of genes, to be inspected later.
Among all cellular functions significantly enriched with genes differentially expressed among selective regimes, I will focus
on those that are most significant, and detailed enough. For example, I skip terms like *transmembrane transport*, *proteolysis*
or *protein phosphorylation*, because they are too general. Sections below are named for the functional categories that I think
are worth discussing. I will try to contextualize the results. My entry point to the literature is [GarciaRoger2019].

I need the results of the differential expression analysis (`2020-01-08`) to identify the genes annotated with the significant
GO terms, and to assess the direction of gene expression change between selective regimes.

```{r setup, message=FALSE}
library(variancePartition)
library(GO.db)
library(ggplot2)
library(tidyr)
library(reactable)
ANNOTATION <- '../2019-07-26/genes/annotation.txt'
EXPRESSION <- '../2020-01-08/genes.RData'   # Includes the variance partition and the mixed models fitted with dream()
```

The annotation table includes only the most specific terms that can be assigned to a
gene. The more general, ancestor terms are assumed, of course, but not explicitly
mentioned in the annotation. For the purpose of retrieving all genes annotated to a
term, I need to expand the annotation, to make ancestor terms explicit. It took me a
while to realize that I can do this with the `unlist()` function:

```{r data}
annotation <- read.table(ANNOTATION, col.names = c('gene', 'GOterms'), colClasses = c('character', 'character'))
head(annotation)
annotation.list <- strsplit(annotation$GOterms, split='|', fixed=TRUE)
names(annotation.list) <- annotation$gene
# append can only join two vectors at a time. It works with lists!
allAncestors <- append(as.list(GOBPANCESTOR),
                       append(as.list(GOMFANCESTOR),
                              as.list(GOCCANCESTOR)))

fullAnnotation <- lapply(annotation.list,
                         FUN = function(x){
                           unique(append(x,
                                         unlist(allAncestors[x], use.names=FALSE)))
                         })
load(EXPRESSION)
ls()
head(varPart)
```

# Biological processes
## Nucleotide-excision repair
This is term **GO:0006289**. It is among the most significantly enriched terms, with only 13 genes annotated.

```{r NER}
goterm <- 'GO:0006289'
genes <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
# Not all genes annotated have been used in expression analysis
genes <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes,])

topTable(fitmm, coef='regime', number=length(fitmm$F))[genes,]

z <- as.data.frame(vobjDream$E[genes,])
z$gene <- row.names(z)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor(NA, levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
GE[GE$sample %in% c("X3A_S9", "X3C_S11", "X5A_S2", "X5C_S4", "X6A_S3", "X6C_S10"), 'regime'] <- 'regular'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

## G protein-coupled receptor signaing pathway, and signal transduction

```{r Gprotein}
goterm <- 'GO:0007186'
genes <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes <- genes[genes %in% row.names(varPart)]
# I subset varPart and at the same time order the selected rows by 'regime', and the columns in the order desired.
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])

z <- as.data.frame(vobjDream$E[genes,])
z$gene <- row.names(z)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor(NA, levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
GE[GE$sample %in% c("X3A_S9", "X3C_S11", "X5A_S2", "X5C_S4", "X6A_S3", "X6C_S10"), 'regime'] <- 'regular'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() + ggtitle(Term(GOTERM[goterm]))
```

Among the `r length(genes)` genes in this category, `r sum(topTable(fitmm, coef='regime', number=length(fitmm$F))[genes, 'adj.P.Val'] <= 0.1)`
have an adjusted p-value lower or equal to 0.1. There are too many genes annotated with this function to plot all their stratified expression levels.
I select the most differentially expressed ones.

```{r G2, fig.width=15, fig.height=20}
z <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z <- z[row.names(z) %in% genes, ]   # gene order is preserved:
reactable(z)

genes <- head(row.names(z), n=50)
GE <- GE[GE$gene %in% genes,]
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() +
      facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

Among the top `r length(genes)` genes analysed from the G protein-coupled receptor signaling pathway,
`r sum(topTable(fitmm, coef='regime', number=length(fitmm$F))[genes, 'logFC'] < 0)`
have a reduced expression level in the unpredictable environment, and
`r sum(topTable(fitmm, coef='regime', number=length(fitmm$F))[genes, 'logFC'] > 0)`
show an increased expression level.

## Cell-matrix adhesion

```{r adhesion}
goterm <- 'GO:0007160'
genes <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes,])

reactable(topTable(fitmm, coef='regime', number=length(fitmm$F))[genes,])  # not ordered by p-value

z <- as.data.frame(vobjDream$E[genes,])
z$gene <- row.names(z)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

Among the top `r length(genes)` genes involved in `r Term(GOTERM[goterm])`, 
`r sum(topTable(fitmm, coef='regime', number=length(fitmm$F))[genes, 'logFC'] < 0)`
have a reduced expression level in the unpredictable environment, and
`r sum(topTable(fitmm, coef='regime', number=length(fitmm$F))[genes, 'logFC'] > 0)`
show an increased expression level.

## Potassium ion transport, carboxylic acid transport, and ion transmembrane transport
Several functions related to ion transport are significant. I wonder to what extent their significance is not
dependent on a common subset of genes.

```{r independence}
goterm_K    <- 'GO:0006813'
goterm_COOH <- 'GO:0046942'
goterm_Sulf <- 'GO:0008272'
goterm_Nucl <- 'GO:1901642'
goterm_TM   <- 'GO:0034220'

genes_K    <- names(fullAnnotation[grep(goterm_K,    fullAnnotation, fixed=TRUE)])
genes_COOH <- names(fullAnnotation[grep(goterm_COOH, fullAnnotation, fixed=TRUE)])
genes_Sulf <- names(fullAnnotation[grep(goterm_Sulf, fullAnnotation, fixed=TRUE)])
genes_Nucl <- names(fullAnnotation[grep(goterm_Nucl, fullAnnotation, fixed=TRUE)])
genes_TM   <- names(fullAnnotation[grep(goterm_TM,   fullAnnotation, fixed=TRUE)])

vennDiagram(vennCounts(cbind(Potassium     = row.names(varPart) %in% genes_K,
                             Carboxilic    = row.names(varPart) %in% genes_COOH,
                             Sulfate       = row.names(varPart) %in% genes_Sulf,
                             Nucleoside    = row.names(varPart) %in% genes_Nucl,
                             Transmembrane = row.names(varPart) %in% genes_TM)))
```

There is hardly any overlap among genes involved in the transport of these substances. The only remarkable overlap
happens between `r Term(GOTERM['GO:0006813'])` and `r Term(GOTERM['GO:0034220'])`. But, still the significance of
each term should be quite independent of the significance of any other term. This was expected, because the analysis
of functional enrichment applied algorithms to account for the dependence structure among GO terms. I analyse them
separately, below.

### Potassium ion transport

```{r potassiumBars, fig.height=10}
goterm <- goterm_K
genes <- genes_K     # defined above
genes <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])
#plotPercentBars(varPart[genes,])
```

```{r potassiumTab}
# I can't keep from using reactable, even if it does not display in the online version...
reactable(topTable(fitmm, coef='regime', number=length(fitmm$F))[genes,],
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))  # not ordered by p-value, just click on the column header to sort.
```

```{r potassiumPlot, fig.width=15, fig.height=20}
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F), sort.by='logFC')
z2 <- z2[row.names(z2) %in% genes,]   # I want to preserve order by logFC
z <- z[row.names(z2),]
# Defining z$gene as a factor with ordered levels will make ggplot order the genes in ascending
# logFC order.
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

### Ion transmembrane transport

```{r ionBars, fig.height=20}
goterm <- goterm_TM
genes <- genes_TM     # defined above.
genes <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])
```

```{r ionTab}
z <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z <- z[row.names(z) %in% genes,]   # still ordered by p-value!
reactable(z,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
```

```{r ionPlot, fig.width=15, fig.height=15}
# To make plots visible, I plot only the most significant terms, say at FDR=0.1.
# Again, I want to select and sort the rows of 'z' at the same time.
genes <- row.names(z)[z$adj.P.Val <= 0.1]
z2 <- z[genes[order(z[genes,'logFC'])],]
# I rename it as z2 to be consistent with previous chunks. This is part of the topTable,
# used only for sorting genes in the expression matrix, named z. Excuse the recycling
# of variable names.

z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
# Defining z$gene as a factor with ordered levels will make ggplot order the genes in ascending
# logFC order.
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

### Nucleoside transmembrane transport

```{r nucleoside}
goterm <- goterm_Nucl
genes  <- genes_Nucl
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])

z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
# I wanted the table to be ordered by FDR, but I need it now ordered by logFC:
z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F), resort.by='logFC')
z2 <- z2[row.names(z2) %in% genes,]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

### Carboxylic acid transport

```{r COOH, fig.height=10}
goterm <- goterm_COOH
genes  <- genes_COOH
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])

z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
# I wanted the table to be ordered by FDR, but I need it now ordered by logFC:
z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F), resort.by='logFC')
z2 <- z2[row.names(z2) %in% genes,]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

## Trehalose biosynthetic process

```{r trehalose}
goterm <- 'GO:0005992'
genes <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])
z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
# It's actually faster to re-order z2 after having selected the rows.
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

## Cell motility, cillium movement and cillium assembly

```{r cillia}
goterm_cellmoti <- 'GO:0048870'   # cell motility
goterm_movement <- 'GO:0003341'   # cillium movement
goterm_assembly <- 'GO:0060271'   # cillium assembly

genes_cellmoti <- names(fullAnnotation[grep(goterm_cellmoti, fullAnnotation, fixed=TRUE)])
genes_movement <- names(fullAnnotation[grep(goterm_movement, fullAnnotation, fixed=TRUE)])
genes_assembly <- names(fullAnnotation[grep(goterm_assembly, fullAnnotation, fixed=TRUE)])

vennDiagram(vennCounts(cbind('Cell motility'    = row.names(varPart) %in% genes_cellmoti,
                             'Cillium movement' = row.names(varPart) %in% genes_movement,
                             'Cillium assembly' = row.names(varPart) %in% genes_assembly)))
```

### Cell motility

```{r cellmoti}
goterm <- goterm_cellmoti
genes  <- genes_cellmoti
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])

z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
# It's actually faster to re-order z2 after having selected the rows.
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

### Cillium movement

```{r movement}
goterm <- goterm_movement
genes  <- genes_movement
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])

z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
# It's actually faster to re-order z2 after having selected the rows.
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) + geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

### Cillium assembly

```{r assembly, fig.heigth=15}
goterm <- goterm_assembly
genes  <- genes_assembly
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])

z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
# It's actually faster to re-order z2 after having selected the rows.
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) +
  geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

## Proteasome-mediated ubiquitin-dependent protein catabolic process

```{r proteasome, fig.heigth=15, fig.width=15}
goterm <- 'GO:0043161'
genes  <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])

z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
# It's actually faster to re-order z2 after having selected the rows.
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) +
  geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

## Response to oxidative stress

```{r oxiestres}
goterm <- 'GO:0006979'
genes  <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])

z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
# It's actually faster to re-order z2 after having selected the rows.
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) +
  geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

# Molecular functions
## Motor activity

```{r motorBars, fig.height=15}
goterm <- 'GO:0003774'
genes  <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])],
                        c('regime', 'treatment', 'population', 'Residuals')])
```

```{r motorTab}
z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
```

```{r motorPlot}
# I limit the number of genes.
z2 <- z2[z2$adj.P.Val <= 0.1,]
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) +
  geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

## Peptidase activities: metallocarboxypeptidase activity, metalloendopeptidase activity, serine-type endopeptidase activity and calcium-dependent cysteine-type endopeptidase activity

```{r peptidase}
goterm_metallocarboxy <- 'GO:0004181'
goterm_metalloendopep <- 'GO:0004222'
goterm_serinetypeendo <- 'GO:0004252'
goterm_calciumcystein <- 'GO:0004198'

genes_metallocarboxy <- names(fullAnnotation[grep(goterm_metallocarboxy, fullAnnotation, fixed=TRUE)])
genes_metalloendopep <- names(fullAnnotation[grep(goterm_metalloendopep, fullAnnotation, fixed=TRUE)])
genes_serinetypeendo <- names(fullAnnotation[grep(goterm_serinetypeendo, fullAnnotation, fixed=TRUE)])
genes_calciumcystein <- names(fullAnnotation[grep(goterm_calciumcystein, fullAnnotation, fixed=TRUE)])

vennDiagram(vennCounts(cbind(Metallocarboxy. = row.names(varPart) %in% genes_metallocarboxy,
                             Metalloendo.    = row.names(varPart) %in% genes_metalloendopep,
                             Serine_type = row.names(varPart) %in% genes_serinetypeendo,
                             Cystein_type = row.names(varPart) %in% genes_calciumcystein)))
```

### Metallocarboxypeptidase activity

```{r metallocarboxypeptidase}
goterm <- 'GO:0004181'
Term(GOTERM[goterm])
genes  <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])

z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
# It's actually faster to re-order z2 after having selected the rows.
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) +
  geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

### Serine-type endopeptidase activity

```{r serineBars, fig.height=15}
goterm <- 'GO:0004252'
Term(GOTERM[goterm])
genes  <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])
```

```{r serineTab}
z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
```

```{r serinePlot, fig.height=10, fig.width=10}
z2 <- z2[z2$adj.P.Val <= 0.1,]     # only significantly regulated genes
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) +
  geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

### Calcium-dependent cysteine-type endopeptidase activity

```{r calciumBars, fig.height=15}
goterm <- 'GO:0004198'
Term(GOTERM[goterm])
genes  <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])
```

```{r calciumTab}
z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
```

```{r calciumPlot, fig.height=10, fig.width=10}
z2 <- z2[z2$adj.P.Val <= 0.13,]     # only significantly regulated genes
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) +
  geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

### Metalloendopeptidase activity

```{r metalloendoBars, fig.height=15}
goterm <- 'GO:0004222'
Term(GOTERM[goterm])
genes  <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])
```

```{r metalloendoTab}
z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
```

```{r metalloendoPlot, fig.height=10, fig.width=10}
z2 <- z2[z2$adj.P.Val <= 0.11,]     # only significantly regulated genes
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) +
  geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

## oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced ascorbate as one donor, and incorporation of one atom of oxygen

```{r oxidoreductaseBars}
goterm <- 'GO:0016715'
Term(GOTERM[goterm])
genes  <- names(fullAnnotation[grep(goterm, fullAnnotation, fixed=TRUE)])
genes  <- genes[genes %in% row.names(varPart)]
plotPercentBars(varPart[genes[order(-varPart[genes, 'regime'])], c('regime', 'treatment', 'population', 'Residuals')])
```

```{r oxidoreductaseTab}
z2 <- topTable(fitmm, coef='regime', number=length(fitmm$F))
z2 <- z2[row.names(z2) %in% genes,]   # still ordered by p-value!
reactable(z2,
          columns = list(
            logFC   = colDef(format=colFormat(digits=3)),
            AveExpr = colDef(format=colFormat(digits=3)),
            t       = colDef(format=colFormat(digits=3)),
            P.Value = colDef(format=colFormat(digits=3)),
            adj.P.Val = colDef (format=colFormat(digits=3)),
            z.std   = colDef(format=colFormat(digits=3))
          ))
```

```{r oxidoreductasePlot, fig.height=10, fig.width=10}
#z2 <- z2[z2$adj.P.Val <= 0.11,]     # only significantly regulated genes
z2 <- z2[order(z2$logFC),]
z  <- as.data.frame(vobjDream$E)   # this is the whole expression matrix
z <- z[row.names(z2),]
z$gene <- factor(row.names(z), levels=row.names(z2), ordered=TRUE)
GE <- pivot_longer(z, cols=c(1:12), names_to='sample', values_to='expression')
GE$regime <- factor('regular', levels=c('regular', 'random'))
GE[GE$sample %in% c("X1A_S8", "X1C_S1", "X2A_S7", "X2C_S5", "X4A_S6", "X4C_S12"), 'regime'] <- 'random'
ggplot(data=GE, mapping=aes(x=regime, y=expression, color=regime)) +
  geom_boxplot() + facet_wrap(~gene) + ggtitle(Term(GOTERM[goterm]))
```

# Session info

# References
